### PR TITLE
Routing improvements

### PR DESCRIPTION
**Pull request by naumenkogs**
A bunch of tiny fixes and introduction of 2 fields: htlc_maximum_msat (on channel update) and channel capacity, which would be later used in #646

### CODE REVIEW

File: lightning/src/ln/channel.rs [hunk 1]
```diff
@@ -3127,6 +3130,11 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 		self.our_htlc_minimum_msat
 	}
 
+	/// Allowed in any state (including after shutdown)
+	pub fn get_our_htlc_maximum_msat(&self) -> Option<u64> {
+		self.our_htlc_maximum_msat
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs:3135]: I think we should always return an htlc_maximum_msat by looking at the channel value (and never returning more than, say, 90% of it whether it was set in an option or not).

Comment (naumenkogs) [reply lightning/src/ln/channel.rs]: I implemented this, but I'm not sure if I understood you 100%. Please double-check.

**Comment by naumenkogs** [comment]
@TheBlueMatt just noticed the issue with fuzzing, do you have an idea of what's wrong here? I'm going through the fuzz logs  and it's not particularly helpful.

File: lightning/src/util/config.rs [hunk 2]
```diff
@@ -59,6 +59,14 @@ pub struct ChannelHandshakeConfig {
 	/// Default value: 1. If the value is less than 1, it is ignored and set to 1, as is required
 	/// by the protocol.
 	pub our_htlc_minimum_msat: u64,
+	/// Set to the largest value HTLC we will accept to process.
```
Comment (TheBlueMatt) [inline lightning/src/util/config.rs:62]: Hmm, I don't think this makes sense - we don't actually *enforce* this, so suggesting to users that it is a "real option" seems strange to me. Nor can we enforce this securely, since our counterparty won't enforce it on inbound HTLCs.

File: lightning/src/ln/channel.rs [hunk 3]
```diff
@@ -3127,6 +3130,15 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 		self.our_htlc_minimum_msat
 	}
 
+	/// Allowed in any state (including after shutdown)
+	pub fn get_our_htlc_maximum_msat(&self) -> u64 {
+		let mut result = (self.channel_value_satoshis as f64 * 0.9).round() as u64; // upper bound by channel capacity
+		if let Some(configured_our_htlc_maximum_msat) = self.our_htlc_maximum_msat {
+			result = cmp::min(configured_our_htlc_maximum_msat, result)
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs:3137]: I think we should also take the min with max_htlc_value_in_flight_msat, since obviousl no single HTLC can violate the total in-flight limit.

File: lightning/src/util/config.rs (hunk 2 above)
Comment (naumenkogs) [reply lightning/src/util/config.rs:62]: I'm a little confused about this option then. I guess it is only used to help others with routing via us?
What about this comment:

/// Set to the largest value HTLC which we expect to be routable via our channels.

/// This value is communicated to other nodes at our channel announcements and is
/// supposed to help other nodes to make better routing decisions and reduce failure rate.

Comment (TheBlueMatt) [reply lightning/src/util/config.rs:62]: > I guess it is only used to help others with routing via us?

Yes, I believe that is the intent.

> What about this comment:

I dont think we need to *store* it at all - it can always be calculated on the fly.

Comment (valentinewallace) [reply lightning/src/util/config.rs:62]: FWIW, lnd and (IIUC) c-lightning enforce max HTLC, and offer it as a configurable option for users. So, putting it out there that I think it'd make sense for us to do so as well :P

File: lightning/src/util/config.rs [hunk 4]
```diff
@@ -59,6 +59,14 @@ pub struct ChannelHandshakeConfig {
...
+	///
+	/// If configured, this value is sent to our counterparty on channel-open and we close the channel any time
+	/// our counterparty misbehaves by sending us an HTLC with a value larger than this.
```
Comment (valentinewallace) [inline lightning/src/util/config.rs:65]: I think other impls just fail the HTLC rather than close on the counterparty -- which I'd prefer as well and I believe the spec doesn't require closing on violation.

File: lightning/src/routing/network_graph.rs [hunk 5]
```diff
@@ -1117,7 +1133,12 @@ mod tests {
 
 	#[test]
 	fn handling_channel_update() {
-		let (secp_ctx, net_graph_msg_handler) = create_net_graph_msg_handler();
+		let secp_ctx = Secp256k1::new();
+		let logger: Arc<Logger> = Arc::new(test_utils::TestLogger::new());
+		let chain_monitor = Arc::new(test_utils::TestChainWatcher::new());
+		let net_graph_msg_handler = NetGraphMsgHandler::new(chain_monitor.clone(), Arc::clone(&logger));
+		*chain_monitor.utxo_ret.lock().unwrap() = Err(chaininterface::ChainError::NotSupported);
```
Comment (valentinewallace) [inline lightning/src/routing/network_graph.rs:1140]: comment for why this line is here?

File: lightning/src/ln/channel.rs [hunk 6]
```diff
@@ -3131,8 +3131,15 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 	}
 
 	/// Allowed in any state (including after shutdown)
-	pub fn get_announced_htlc_max_msat(&self) -> Option<u64> {
-		self.our_htlc_maximum_msat
+	pub fn get_announced_htlc_max_msat(&self) -> u64 {
+		let mut result = cmp::min(
+			self.channel_value_satoshis * 9 / 10, // upper bound by capacity
+			Channel::<ChanSigner>::get_our_max_htlc_value_in_flight_msat(self.channel_value_satoshis)
+		);
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:3138]: @TheBlueMatt why not just max_inflight? the 9/10 seems potentially confusing, so just want to make sure it's actually necessary...

File: lightning/src/ln/channel.rs [hunk 7]
```diff
@@ -3131,7 +3131,7 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 	}
 
 	/// Allowed in any state (including after shutdown)
-	pub fn get_our_htlc_maximum_msat(&self) -> Option<u64> {
+	pub fn get_announced_htlc_max_msat(&self) -> Option<u64> {
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs]: nit: change doesn't seem to belong in this commit

File: lightning/src/ln/channel.rs (hunk 6 above)
Comment (naumenkogs) [reply lightning/src/ln/channel.rs:3138]: I think this is just a heuristic to reduce the failure rate in case if the channel was previously used to route a few sats...
It helps with both malicious and (very trivial) non-malicious cases.

If I understand it correctly, I do think it's a useful bound.
I'm not sure for whom this would be confusing, maybe you have a better intuition of lightning users/operators?

File: lightning/src/ln/msgs.rs [hunk 8]
```diff
@@ -1189,22 +1210,30 @@ impl Writeable for UnsignedChannelUpdate {
...
 			chain_hash: Readable::read(r)?,
 			short_channel_id: Readable::read(r)?,
 			timestamp: Readable::read(r)?,
-			flags: Readable::read(r)?,
+			flags: {
+				let flags = Readable::read(r)?;
+				let message_flags = flags >> 8;
+				has_htlc_maximum_msat = (message_flags as i32 & 1) == 1;
+				flags
+			},
```
[the remark this answers is not in the document]
Comment (naumenkogs) [reply lightning/src/ln/msgs.rs:1231]: There are multiple ways to test this, one of them is already [here](https://github.com/naumenkogs/rust-lightning/blob/5f551a8ef5d1b3e5523fbfb961ef03c24e7de009/lightning/src/routing/network_graph.rs#L1249).
If you prefer something else, could you be more specific?

File: lightning/src/util/config.rs (hunk 4 above)
Comment (naumenkogs) [reply lightning/src/util/config.rs:65]: Same being discussed [here](https://github.com/rust-bitcoin/rust-lightning/pull/651#discussion_r449231788): whether to make it configurable and whether to enforce it.

File: lightning/src/util/config.rs (hunk 2 above)
Comment (naumenkogs) [reply lightning/src/util/config.rs:62]: I don't feel like this option is particularly useful for LN node operators, so I'm fine with just dropping it for the sake of less code...

Matt's opinion on your suggestion would be useful here I guess.

Comment (valentinewallace) [reply lightning/src/util/config.rs:62]: That's a fair point, don't feel too strongly so I'm fine to continue on the course of removing it.

**Comment by naumenkogs** [comment]
Ready for review again :)
Would use some help with figuring why fuzzing fails though.

File: lightning/src/ln/channel.rs (hunk 6 above)
Comment (valentinewallace) [reply lightning/src/ln/channel.rs]: I guess I just meant that I was confused by it, lol.

OK, seems like that'd be worth documenting. I'd add a comment along the lines of -- 
"Upper bound our max_htlc by 9/10 of the channel capacity. This protects against the following attack:
If the channel has already routed a few sats, then without this upper bound the attacker may be able to DoS our node by repeatedly attempting and failing to route `our_channel_capacity - 1` to us. Whereas, if we advertise our max_htlc as 9/10 of channel capacity, no one will try to route `our_channel_capacity_amount` via our channels, so an attacker would at least have to pay a nontrivial amount to accomplish the same DoS attack."

**Comment by valentinewallace** [comment]
> Ready for review again :)
> Would use some help with figuring why fuzzing fails though.

I was able to make progress on this by following the instructions here: https://github.com/rust-bitcoin/rust-lightning/tree/master/fuzz#a-fuzz-test-failed-on-travis-what-do-i-do

Looks related to writing out the excess data on an UnsignedChannelUpdate.

File: lightning/src/ln/channel.rs (hunk 6 above)
Comment (naumenkogs) [reply lightning/src/ln/channel.rs]: I actually meant an attacker routing those fews sats in the first place, not failing to route... As you see, the scope may be broad here. As well as non-malicious cases when few sats were routed by honest users.
I would add one sentence, something about amount "slack", but not sure how to word it better.

File: lightning/src/ln/msgs.rs [hunk 9]
```diff
@@ -742,6 +743,26 @@ impl Readable for OptionalField<Script> {
...
+			OptionalField::Absent => {}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for OptionalField<u64> {
+	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
+		let value: u64 = Readable::read(r)?;
```
Comment (naumenkogs) [inline lightning/src/ln/msgs.rs:760]: I think this is why fuzzing fails... Well, at least I see why fuzzer *might* fail here.

`Serialized(Deserialized(UnsignedChannelAnnouncement1)) != UnsignedChannelAnnouncement1` if htlc_maximum_msat is present, but the flag is not set (or vice versa). Because the flag logic is one level above these serializers.
And fuzzers are probably not aware of this logic?
Not sure how to make it work properly.

File: lightning/src/ln/channel.rs (hunk 6 above)
Comment (valentinewallace) [reply lightning/src/ln/channel.rs]: Gotcha -- I'd like to be super explicit about what exactly we're protecting against, to be well-documented for future developers. Let me know what phrasing you think would be best :)

File: lightning/src/ln/msgs.rs [hunk 10]
```diff
@@ -1608,6 +1637,7 @@ mod tests {
 			flags: if direction { 1 } else { 0 } | if disable { 1 << 1 } else { 0 } | if htlc_maximum_msat { 1 << 8 } else { 0 },
 			cltv_expiry_delta: 144,
 			htlc_minimum_msat: 1000000,
+			htlc_maximum_msat: if htlc_maximum_msat { OptionalField::Present(1) } else { OptionalField::Absent },
```
Comment (TheBlueMatt) [inline lightning/src/ln/msgs.rs:1640]: Generally best to use a value which has more than one byte non-0 to test endianness.

File: lightning/src/ln/msgs.rs [hunk 11]
```diff
@@ -1636,6 +1666,7 @@ mod tests {
 		}
 		target_value.append(&mut hex::decode("009000000000000f42400000271000000014").unwrap());
 		if htlc_maximum_msat {
+			target_value.append(&mut hex::decode("0000000000000001").unwrap());
```
Comment (TheBlueMatt) [inline lightning/src/ln/msgs.rs:1669]: This is confusing - the next line now refers exclusively to excess data which is *not* the htlc_maximum_msat. Comment would be fine but separate out the excess data testing from htlc_maximum testing.

Comment (naumenkogs) [reply lightning/src/ln/msgs.rs:1669]: Sure, I just overlooked this intersection.

Comment (naumenkogs) [reply lightning/src/ln/msgs.rs:1669]: I fixed the code. Not sure why the comment is useful, I think it's self-explanatory :)

File: lightning/src/routing/network_graph.rs [hunk 12]
```diff
@@ -658,6 +665,21 @@ impl NetworkGraph {
...
+
+					if let Some(capacity_sats) = channel.capacity_sats {
+						// It's possible channel capacity is available now, although it wasn't available at announcement (so the field is None).
+						// Don't query UTXO set here to reduce DoS risks.
+						if htlc_maximum_msat > capacity_sats * 1000 {
+							return Err(LightningError{err: "htlc_maximum_msat is larger than channel capacity", action: ErrorAction::IgnoreError});
+						}
+					}
+
+					// TODO check that htlc_maximum_msat is less than max_htlc_value_in_flight_msat
```
[the remark this answers is not in the document]
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:681]: The spec says `MUST set this to less than or equal to max_htlc_value_in_flight_msat it received from the peer.`.

I'm somewhat lost here... How can we even get this field (`max_htlc_value_in_flight_msat`)? I see it in OpenChannel/AcceptChannel messages, but those are relevant for our channels, not remote channels (for which we will receive channel updates).

File: lightning/src/ln/msgs.rs [hunk 13]
```diff
@@ -430,6 +430,7 @@ pub(crate) struct UnsignedChannelUpdate {
 	pub(crate) flags: u16,
 	pub(crate) cltv_expiry_delta: u16,
 	pub(crate) htlc_minimum_msat: u64,
+	pub(crate) htlc_maximum_msat: OptionalField<u64>,
```
[the remark this answers is not in the document]
Comment (naumenkogs) [reply lightning/src/ln/msgs.rs:433]: I implemented it, but now fuzzing fails. I believe this is what happens:
1. It generates an UnsignedChannelUpdate with flags=39835 (u16, all flags)
2. It means 155 to store, and 155 as message_flags
3. When deserializing, it can only make message_flags=1, not 155
4. The end-value is (1<<8) | 155, not (155 << 8) | 155

What should we do with this?

File: lightning/src/routing/network_graph.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs:681]: Ohhh, sorry, I'd missed that that field isnt in the routing gossip. No, so what we should be doing is including `max_htlc_value_in_flight` in `ChannelDetails` in channelmanager and then use that via the `first_hops` argument in get_route. No need to bother crossing layers to enforce the rule in the routing graph when we already allow next-hop channels to override the graph completely.

File: lightning/src/ln/msgs.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs:433]: Ah! Yea, so that's fine, just need to change the fuzz target type, though sadly we may need to define a new one. There are various different rules we apply when fuzzing different messages, you can find them at `fuzz/src/msg_targets/utils.rs` - essentially what we want is `test_msg_hole` but instead of forcing the "hole" to be all-0s, we can just let it be anything (well, really either 0 or 1, but just make it generic). Easiest is likely just replace `test_msg_hole` with a parameter on whether the hole has to be 0 or can be anything.

File: lightning/src/routing/network_graph.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs]: Let me re-iterate on your suggestion:
1. Add `max_htlc_value_in_flight` field to `ChannelDetails` struct
2. When calling for `ChannelDetails` by `list_channels`, this value should be filled from the `Channel` record
3. Then a user will be providing these channels to `get_route` via `first_hops` field
4. Inside `get_route`, make sure `max_htlc_value_in_flight` is not violated by the payment we're sending.

But this has absolutely nothing to do with `htlc_maximum_msat`, which was why I initially added this comment to implement a check per the spec?
So you are suggesting to just ignore that spec part, and implement the flow I listed above?

File: lightning/src/ln/msgs.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (naumenkogs) [reply lightning/src/ln/msgs.rs:433]: 1. I don't see where `test_msg_hole` is forced to be all-0s. It already looks like anything (although the comment says it should be 0s). It just seems to check that everything before and after the hole equals.
2. Said that, I just switched the relevant target to `test_msg_hole 108 1`. Now it passes locally, but something is wrong with CI.

File: lightning/src/routing/network_graph.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs]: Yea, definitely belongs in a separate pr, and your understanding is correct. This TODO is just out of place given that context.

File: lightning/src/ln/msgs.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs:433]: Oof. Can you fix the comment there, at least?

File: lightning/src/ln/channelmanager.rs [hunk 14]
```diff
@@ -1186,7 +1186,9 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 							res.extend_from_slice(&byte_utils::be32_to_array(msg.cltv_expiry));
 						}
 						else if code == 0x1000 | 20 {
-							res.extend_from_slice(&byte_utils::be16_to_array(chan_update.contents.flags));
+							let mut flags = chan_update.contents.flags as u16;
+							if let OptionalField::Present(_) = chan_update.contents.htlc_maximum_msat {flags |= 1 << 8};
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:1190]: flags is a u8, I think this diff hunk can be droped.

File: lightning/src/ln/channelmanager.rs [hunk 15]
```diff
@@ -2494,7 +2497,9 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 							let reason = if let Ok(upd) = self.get_channel_update(chan) {
 								onion_utils::build_first_hop_failure_packet(incoming_shared_secret, error_code, &{
 									let mut res = Vec::with_capacity(8 + 128);
-									res.extend_from_slice(&byte_utils::be16_to_array(upd.contents.flags));
+									let mut flags = upd.contents.flags as u16;
+									if let OptionalField::Present(_) = upd.contents.htlc_maximum_msat {flags |= 1 << 8};
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:2501]: Same here.

File: lightning/src/ln/msgs.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (naumenkogs) [reply lightning/src/ln/msgs.rs:433]: done!

File: lightning/src/ln/channelmanager.rs (hunk 14 above)
Comment (naumenkogs) [reply lightning/src/ln/channelmanager.rs:1190]: `chan_update.contents.flags` is u8, so I get this compile error: `expected `u16`, found `u8``
If I switch to `res.push(upd.contents.flags);`, 4 tests fail.

But also, I don't understand why you're suggesting to omit the `message_flags` part of flags?

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1190]: Oh, interesting, I think this is a misreading of the spec (in the original code). See https://github.com/lightningnetwork/lightning-rfc/issues/791 but my read is that we should likely be pushing two zero bytes, not the flags bytes from the update.

Comment (naumenkogs) [reply lightning/src/ln/channelmanager.rs]: Addressed.
